Make fullscreen API use the top layer

nt1m opened the pull request:
#### c96dd4f574c01a761f875ce4222725c44693f2b1
<pre>
Make fullscreen API use the top layer
<a href="https://bugs.webkit.org/show_bug.cgi?id=84798">https://bugs.webkit.org/show_bug.cgi?id=84798</a>
rdar://11313423

Reviewed by Darin Adler.

The fullscreen API should use the top layer instead of the fullscreen stack. Some notes:
- Pushing to the top layer causes visual changes, so we need to update the DOM methods to follow spec order to prevent unintended effects.
Visual changes can&apos;t happen before `willEnterFullscreen` and can&apos;t happen before `didExitFullscreen` when fully entering/exiting fullscreen.

- Since top layer handles visual z-ordering, we can remove old style hacks of clobbering the ancestor chain with special styling, and the associated
containsFullscreenElement node flag.

Spec: <a href="https://fullscreen.spec.whatwg.org">https://fullscreen.spec.whatwg.org</a>

* LayoutTests/fullscreen/full-screen-enter-while-exiting-expected.txt:
Events are now scheduled and dispatched in didExitFullscreen instead of exitFullscreen(), hence the order change.

* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::selectorText const):
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/css/SelectorPseudoClassAndCompatibilityElementMap.in:
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::addPseudoClassType):
Remove temporary backdrop workaround that was landed in <a href="https://commits.webkit.org/256226@main">https://commits.webkit.org/256226@main</a> .

* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const):
* Source/WebCore/css/SelectorCheckerTestFunctions.h:
(WebCore::matchesFullScreenPseudoClass):
(WebCore::matchesFullScreenAncestorPseudoClass):
Rewrite selector checker functions to account for removal of containsFullscreenElement flag.
Since :-webkit-fullscreen-ancestor main usage was for the UA stylesheet, we may be able to remove it in a followup clean up.

(WebCore::matchesFullScreenParentPseudoClass): Deleted. (part of the backdrop workaround)
* Source/WebCore/css/fullscreen.css:
(#if defined(ENABLE_FULLSCREEN_API) &amp;&amp; ENABLE_FULLSCREEN_API):
(:root:-webkit-full-screen-document:not(:-webkit-full-screen)): Only keep overflow: hidden; since top layer does not make the root element unscrollable.
(iframe:-webkit-full-screen): Match the spec.
(:not(:root):-webkit-full-screen::backdrop): Match the spec by making the fullscreen backdrops black.
(:-webkit-full-screen): Deleted. (not needed with top layer)
(:root:-webkit-full-screen-document:not(:-webkit-full-screen), :root:-webkit-full-screen-ancestor): Deleted.
(:-webkit-full-screen-ancestor:not(iframe)): Deleted. (not needed with top layer)
(video:-webkit-full-screen, audio:-webkit-full-screen): Deleted. (to match the spec)
(:-webkit-full-screen-parent::before): Deleted. (fake backdrop workaround)
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::nodeChildrenWillBeRemoved):
(WebCore::Document::nodeWillBeRemoved):
Move element removal handling to `Element::removedFromAncestor()`

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::insertedIntoAncestor):
(WebCore::Element::removedFromAncestor):
(WebCore::Element::setFullscreenFlag):
(WebCore::Element::setIframeFullscreenFlag):
(WebCore::parentCrossingFrameBoundaries): Deleted.
(WebCore::Element::setContainsFullScreenElement): Deleted.
(WebCore::Element::setContainsFullScreenElementOnAncestorsCrossingFrameBoundaries): Deleted.
Removed the containsFullscreenElement flag and replace it with fullscreen flag and iframe fullscreen flag.

* Source/WebCore/dom/Element.h: Ditto
(WebCore::Element::hasFullscreenFlag const):
(WebCore::Element::hasIframeFullscreenFlag const):
(WebCore::Element::containsFullScreenElement const): Deleted.

* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::fullscreenElement const): Rewrite it to iterate through top layer instead of fullscreen stack.
(WebCore::FullscreenManager::requestFullscreenForElement): Removed all fullscreen stack handling. Top layer handling is now in `willEnterFullscreen` since it performs visual changes.
(WebCore::FullscreenManager::cancelFullscreen): Basically exitFullscreen() but always exits everything.
(WebCore::documentsToUnfullscreen): &quot;collecting documents to unfullscreen&quot; algorithm from spec
(WebCore::FullscreenManager::exitFullscreen): Steps 1 - 8 of exit fullscreen algorithm from spec
(WebCore::FullscreenManager::finishExitFullscreen): Steps 12 - 15 of exit fullscreen algorithm from spec
(WebCore::FullscreenManager::willEnterFullscreen): Push documents to top layer, add fullscreen flag &amp; emit events (Step 12 of request fullscreen in spec)
(WebCore::FullscreenManager::didExitFullscreen): Everything else that is needed when fully exiting fullscreen.
(WebCore::FullscreenManager::exitRemovedFullscreenElementIfNeeded): &quot;removing steps&quot; for element from spec (except for top layer removal which is already done later on in Element::removedFromAncestor)
(WebCore::FullscreenManager::isSimpleFullscreenDocument const): &quot;Simple fullscreen document&quot; as defined by the spec, e.g. a document with only one fullscreen element in the top layer.

(WebCore::FullscreenManager::adjustFullscreenElementOnNodeRemoval): Deleted. Replaced by exitRemovedFullscreenElementIfNeeded called from Element::removedFromAncestor().

(WebCore::FullscreenManager::clearFullscreenElementStack): Deleted.
(WebCore::FullscreenManager::popFullscreenElementStack): Deleted.
(WebCore::FullscreenManager::pushFullscreenElementStack): Deleted.
(WebCore::FullscreenManager::clear):
Remove fullscreen stack remainders.

* Source/WebCore/dom/FullscreenManager.h:

* Source/WebCore/dom/Node.h:
Add IsFullscreen/IsIframeFullscreen flags, remove ContainsFullScreenElement.
Shuffle flags around, since Bit 31 is actually not usable (causes integer overflow when using it).

Canonical link: <a href="https://commits.webkit.org/257456@main">https://commits.webkit.org/257456@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2dad342434ac3e7290cfad31992d4d5229f549d6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99088 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8291 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32224 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108482 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/168724 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103051 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8836 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/85641 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/91591 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/106426 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104813 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/6684 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/90260 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/91591 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/88506 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/21605 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/91591 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2185 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/23121 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2077 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/85641 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5135 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/7040 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42597 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3493 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/Node.h:
```diff
@@ -568,29 +568,27 @@ class Node : public EventTarget {
-        NeedsUpdateQueryContainerDependentStyle = 1 << 29,
-        HasBeenInUserAgentShadowTree = 1 << 30,
-
-        // Bit 31 is free.
```
This bit is still free, right?

nt1m replied:
I remember getting an integer overflow when trying to use it ~2 months ago, but perhaps that's just a bug with the existing code.

darinadler replied:
Pretty sure the bit is still free; doesn’t seem right to remove the comment.

darinadler commented on Source/WebCore/css/SelectorCheckerTestFunctions.h:
```diff
@@ -422,17 +422,12 @@ ALWAYS_INLINE bool matchesFullScreenAnimatingFullScreenTransitionPseudoClass(con
+    auto* currentFullscreenElement = element.document().fullscreenManager().currentFullscreenElement();
+    if (!currentFullscreenElement || &element == currentFullscreenElement)
+        return false;
+    return element.document().fullscreenManager().currentFullscreenElement()->isDescendantOrShadowDescendantOf(element);
```
Should use the local variable `currentFullscreenElement` here.

darinadler commented on Source/WebCore/css/SelectorCheckerTestFunctions.h:
```diff
@@ -422,17 +422,12 @@ ALWAYS_INLINE bool matchesFullScreenAnimatingFullScreenTransitionPseudoClass(con
 {
-    return element.containsFullScreenElement();
+    auto* currentFullscreenElement = element.document().fullscreenManager().currentFullscreenElement();
+    if (!currentFullscreenElement || &element == currentFullscreenElement)
```
The `&element == currentFullscreenElement` check should not be needed here. `isDescendantOrShadowDescendantOf` will return false without an explicit check. I would write:
```
return currentFullscreenElement && currentFullscreenElement-> isDescendantOrShadowDescendantOf(element);
```

darinadler commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -4181,20 +4167,22 @@ void Element::requestFullscreen(FullscreenOptions&&, RefPtr<DeferredPromise>&& p
+        setNodeFlag(NodeFlag::IsFullscreen);
+    else {
+        clearNodeFlag(NodeFlag::IsFullscreen);
+        clearNodeFlag(NodeFlag::IsIframeFullscreen);
```
Is it really OK to leave this flag alone when setting fullscreen to true? I don’t really like the side effect when clearing the flag to clear another flag.

nt1m replied:
Yes, the iframe fullscreen flag should only be set when a iframe was explicitly fullscreened, not implicitly fullscreened by its descendants.

darinadler replied:
Then could this _clear_ `NodeFlag::IsIframeFullscreen`? I still think it’s surprising that this clears the other flag, and only does it as a side effect of clearing the one. If this is going to touch the other as a side effect, perhaps we should model this as three states: Not full screen, non-iframe full screen, iframe full screen.

nt1m replied:
Sure, I can clear it.

darinadler commented on Source/WebCore/dom/Node.h:
```diff
@@ -568,29 +568,27 @@ class Node : public EventTarget {
 #if ENABLE(FULLSCREEN_API)
-        ContainsFullScreenElement = 1 << 24,
+        IsFullscreen = 1 << 29,
+        IsIframeFullscreen = 1 << 30,
```
Not so fond of making iframe a word here. I would probably have used `IFrame` instead of `Iframe`.

darinadler commented on Source/WebCore/dom/FullscreenManager.h:
```diff
@@ -77,7 +78,10 @@ class FullscreenManager final : public CanMakeWeakPtr<FullscreenManager> {
 
-    void adjustFullscreenElementOnNodeRemoval(Node&, Document::NodeRemoval = Document::NodeRemoval::Node);
+    enum class ExitMode : bool { Resize, NoResize };
+    void finishExitFullscreen(Document& currentDoc, ExitMode);
```
Please use the word `document` rather than the abbreviation `doc`. But also, does this argument really need a name? Not sure the word current really disambiguates which document should be passed.

darinadler commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -588,23 +585,17 @@ void FullscreenManager::dispatchFullscreenChangeOrErrorEvent(Deque<GCReachableRe
-        cancelFullscreen();
-    }
+    auto fullscreenElement = fullscreenOrPendingElement();
+    if (fullscreenElement && fullscreenElement == &element) {
```
No need for the null check here. A null pointer will never be `== &element`. So:
```
if (fullscreenOrPendingElement() == &element)
```

darinadler commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -301,84 +265,61 @@ void FullscreenManager::cancelFullscreen()
+
+        // This triggers finishExitFullscreen with ExitMode::Resize, which fully exits the document.
+        if (auto* fullscreenElement = topDocument.fullscreenManager().fullscreenElement())
+            page()->chrome().client().exitFullScreenForElement(fullscreenElement);
```
Seems like `exitFullScreenForElement` should take a reference, not a pointer. Something to refine in the future, I guess.

darinadler commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -301,84 +265,61 @@ void FullscreenManager::cancelFullscreen()
 }
 
-void FullscreenManager::exitFullscreen()
+static Vector<RefPtr<Document>> documentsToUnfullscreen(Document& doc)
```
I suggest making the return type `Vector<Ref<Document>>`.

Also, please use the word `document` rather than the abbreviation `doc` for the argument. Also, unclear what the document argument is since the function returns documents. Maybe the argument name should be `firstDocument` or `startDocument` or `elementDocument`.

darinadler commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -301,84 +265,61 @@ void FullscreenManager::cancelFullscreen()
-        INFO_LOG(LOGIDENTIFIER, "Cancelling pending fullscreen request.");
-        m_pendingFullscreenElement = nullptr;
-        return;
+    Vector<RefPtr<Document>> docs { RefPtr { &doc } };
```
Please use `documents` rather than the abbreviation `docs`.

darinadler commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -301,84 +265,61 @@ void FullscreenManager::cancelFullscreen()
-        return;
+    Vector<RefPtr<Document>> docs { RefPtr { &doc } };
+    while (true) {
+        auto lastDoc = docs.last();
```
And `lastDocument` rather than `lastDoc`.

darinadler commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -301,84 +265,61 @@ void FullscreenManager::cancelFullscreen()
+        ASSERT(lastDoc->fullscreenManager().fullscreenElement());
+        if (!lastDoc->fullscreenManager().isSimpleFullscreenDocument())
+            break;
+        auto parentFrame = lastDoc->frame()->ownerElement();
```
What guarantees that `lastDocument->frame()` is non-null here?

A local variable that is an element should probably not be named frame. Maybe `frameOwner` instead of `parentFrame`?

darinadler commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -301,84 +265,61 @@ void FullscreenManager::cancelFullscreen()
+        if (!lastDoc->fullscreenManager().isSimpleFullscreenDocument())
+            break;
+        auto parentFrame = lastDoc->frame()->ownerElement();
+        if (!parentFrame || (parentFrame && parentFrame->hasIframeFullscreenFlag()))
```
No need to doubly check for null. I would write it more like this:
```
auto frameOwner = parentFrame->ownerElement();
if (!frameOwner || frameOwner->hasIframeFullscreenFlag())
    break;
```

darinadler commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -301,84 +265,61 @@ void FullscreenManager::cancelFullscreen()
-        if (localFrame->document()->fullscreenManager().fullscreenElement())
-            descendants.prepend(localFrame->document());
-    }
+    auto* currentDoc = &document();
```
Please `currentDocument` instead of `currentDoc`. But also I am not sure that "current" is the right word.

nt1m replied:
This matched spec terminology, but I changed it to `exitingDocument` which is a bit more accurate.

darinadler commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -301,84 +265,61 @@ void FullscreenManager::cancelFullscreen()
-    }
+    auto* currentDoc = &document();
+    auto mode = ExitMode::NoResize;
+    auto exitDocs = documentsToUnfullscreen(*currentDoc);
```
Please use `exitDocuments` instead of `exitDocs`. But also unclear what "exit documents" are.

darinadler commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -301,84 +265,61 @@ void FullscreenManager::cancelFullscreen()
+    if (exitDocs.contains(&topDocument()) && topDocument().fullscreenManager().isSimpleFullscreenDocument()) {
+        mode = ExitMode::Resize;
+        currentDoc = &topDocument();
     }
```
I think we should not call topDocument() three times. It’s not free to compute. So we can use a local variable.

darinadler commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -301,84 +265,61 @@ void FullscreenManager::cancelFullscreen()
-
-        // 2. Queue a task to fire an event named fullscreenchange with its bubbles attribute set to true
-        // on doc.
+    auto currentDocFullscreenElement = currentDoc->fullscreenManager().fullscreenElement();
```
I think we should just call this local `fullscreenElement`. Let’s not use the abbreviation `doc` here.

darinadler commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -392,28 +333,71 @@ void FullscreenManager::exitFullscreen()
 
-        // Otherwise, notify the chrome of the new full screen element.
-        m_pendingExitFullscreen = false;
+void FullscreenManager::finishExitFullscreen(Document& currentDoc, ExitMode mode)
```
I’ll stop commenting about every use of `doc`, but there are more here.

darinadler reviewed (approved):
This looks good to me. I’m still not sure we have enough testing of this, we’d want tests that fail if any one line of this code was wrong. Regardless, r=me

nt1m committed:
Make fullscreen API use the top layer
https://bugs.webkit.org/show_bug.cgi?id=84798
rdar://11313423

Reviewed by Darin Adler.

The fullscreen API should use the top layer instead of the fullscreen stack. Some notes:
- Pushing to the top layer causes visual changes, so we need to update the DOM methods to follow spec order to prevent unintended effects.
Visual changes can't happen before `willEnterFullscreen` and can't happen before `didExitFullscreen` when fully entering/exiting fullscreen.

- Since top layer handles visual z-ordering, we can remove old style hacks of clobbering the ancestor chain with special styling, and the associated
containsFullscreenElement node flag.

Spec: https://fullscreen.spec.whatwg.org

* LayoutTests/fullscreen/full-screen-enter-while-exiting-expected.txt:
Events are now scheduled and dispatched in didExitFullscreen instead of exitFullscreen(), hence the order change.

* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::selectorText const):
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/css/SelectorPseudoClassAndCompatibilityElementMap.in:
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::addPseudoClassType):
Remove temporary backdrop workaround that was landed in https://commits.webkit.org/256226@main .

* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const):
* Source/WebCore/css/SelectorCheckerTestFunctions.h:
(WebCore::matchesFullScreenPseudoClass):
(WebCore::matchesFullScreenAncestorPseudoClass):
Rewrite selector checker functions to account for removal of containsFullscreenElement flag.
Since :-webkit-fullscreen-ancestor main usage was for the UA stylesheet, we may be able to remove it in a followup clean up.

(WebCore::matchesFullScreenParentPseudoClass): Deleted. (part of the backdrop workaround)
* Source/WebCore/css/fullscreen.css:
(#if defined(ENABLE_FULLSCREEN_API) && ENABLE_FULLSCREEN_API):
(:root:-webkit-full-screen-document:not(:-webkit-full-screen)): Only keep overflow: hidden; since top layer does not make the root element unscrollable.
(iframe:-webkit-full-screen): Match the spec.
(:not(:root):-webkit-full-screen::backdrop): Match the spec by making the fullscreen backdrops black.
(:-webkit-full-screen): Deleted. (not needed with top layer)
(:root:-webkit-full-screen-document:not(:-webkit-full-screen), :root:-webkit-full-screen-ancestor): Deleted.
(:-webkit-full-screen-ancestor:not(iframe)): Deleted. (not needed with top layer)
(video:-webkit-full-screen, audio:-webkit-full-screen): Deleted. (to match the spec)
(:-webkit-full-screen-parent::before): Deleted. (fake backdrop workaround)
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::nodeChildrenWillBeRemoved):
(WebCore::Document::nodeWillBeRemoved):
Move element removal handling to `Element::removedFromAncestor()`

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::insertedIntoAncestor):
(WebCore::Element::removedFromAncestor):
(WebCore::Element::setFullscreenFlag):
(WebCore::Element::setIframeFullscreenFlag):
(WebCore::parentCrossingFrameBoundaries): Deleted.
(WebCore::Element::setContainsFullScreenElement): Deleted.
(WebCore::Element::setContainsFullScreenElementOnAncestorsCrossingFrameBoundaries): Deleted.
Removed the containsFullscreenElement flag and replace it with fullscreen flag and iframe fullscreen flag.

* Source/WebCore/dom/Element.h: Ditto
(WebCore::Element::hasFullscreenFlag const):
(WebCore::Element::hasIframeFullscreenFlag const):
(WebCore::Element::containsFullScreenElement const): Deleted.

* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::fullscreenElement const): Rewrite it to iterate through top layer instead of fullscreen stack.
(WebCore::FullscreenManager::requestFullscreenForElement): Removed all fullscreen stack handling. Top layer handling is now in `willEnterFullscreen` since it performs visual changes.
(WebCore::FullscreenManager::cancelFullscreen): Basically exitFullscreen() but always exits everything.
(WebCore::documentsToUnfullscreen): "collecting documents to unfullscreen" algorithm from spec
(WebCore::FullscreenManager::exitFullscreen): Steps 1 - 8 of exit fullscreen algorithm from spec
(WebCore::FullscreenManager::finishExitFullscreen): Steps 12 - 15 of exit fullscreen algorithm from spec
(WebCore::FullscreenManager::willEnterFullscreen): Push documents to top layer, add fullscreen flag & emit events (Step 12 of request fullscreen in spec)
(WebCore::FullscreenManager::didExitFullscreen): Everything else that is needed when fully exiting fullscreen.
(WebCore::FullscreenManager::exitRemovedFullscreenElementIfNeeded): "removing steps" for element from spec (except for top layer removal which is already done later on in Element::removedFromAncestor)
(WebCore::FullscreenManager::isSimpleFullscreenDocument const): "Simple fullscreen document" as defined by the spec, e.g. a document with only one fullscreen element in the top layer.

(WebCore::FullscreenManager::adjustFullscreenElementOnNodeRemoval): Deleted. Replaced by exitRemovedFullscreenElementIfNeeded called from Element::removedFromAncestor().

(WebCore::FullscreenManager::clearFullscreenElementStack): Deleted.
(WebCore::FullscreenManager::popFullscreenElementStack): Deleted.
(WebCore::FullscreenManager::pushFullscreenElementStack): Deleted.
(WebCore::FullscreenManager::clear):
Remove fullscreen stack remainders.

* Source/WebCore/dom/FullscreenManager.h:

* Source/WebCore/dom/Node.h:
Add IsFullscreen/IsIframeFullscreen flags, remove ContainsFullScreenElement.
Shuffle flags around, since Bit 31 is actually not usable (causes integer overflow when using it).

Canonical link: https://commits.webkit.org/257456@main
